Flaky: Flux end to end tests

I got this red run on a PR: https://circleci.com/gh/influxdata/influxdb/36696

```
--- FAIL: TestFluxEndToEnd (23.09s)
    --- PASS: TestFluxEndToEnd/aggregate_empty_window.flux (0.42s)
    --- FAIL: TestFluxEndToEnd/aggregate_fill_window.flux (0.61s)
    --- FAIL: TestFluxEndToEnd/aggregate_window.flux (0.53s)
    --- FAIL: TestFluxEndToEnd/limit_offset.flux (0.58s)
    --- FAIL: TestFluxEndToEnd/lowestCurrent.flux (0.53s)
```

I've attached the build file.

[build_36696_step_108_container_0.txt](https://github.com/influxdata/influxdb/files/3007865/build_36696_step_108_container_0.txt)

--- a/query/stdlib/influxdata/influxdb/source.go
+++ b/query/stdlib/influxdata/influxdb/source.go
@@ -55,17 +55,7 @@ func (s *Source) Metadata() flux.Metadata {
 
 func (s *Source) processTables(ctx context.Context, tables TableIterator, watermark execute.Time) error {
 	err := tables.Do(func(tbl flux.Table) error {
-		for _, t := range s.ts {
-			if err := t.Process(s.id, tbl); err != nil {
-				return err
-			}
-			//TODO(nathanielc): Also add mechanism to send UpdateProcessingTime calls, when no data is arriving.
-			// This is probably not needed for this source, but other sources should do so.
-			if err := t.UpdateProcessingTime(s.id, execute.Now()); err != nil {
-				return err
-			}
-		}
-		return nil
+		return s.processTable(ctx, tbl)
 	})
 	if err != nil {
 		return err
@@ -78,6 +68,30 @@ func (s *Source) processTables(ctx context.Context, tables TableIterator, waterm
 
 	for _, t := range s.ts {
 		if err := t.UpdateWatermark(s.id, watermark); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Source) processTable(ctx context.Context, tbl flux.Table) error {
+	if len(s.ts) == 0 {
+		tbl.Done()
+		return nil
+	} else if len(s.ts) == 1 {
+		return s.ts[0].Process(s.id, tbl)
+	}
+
+	// There is more than one transformation so we need to
+	// copy the table for each transformation.
+	bufTable, err := execute.CopyTable(tbl)
+	if err != nil {
+		return err
+	}
+	defer bufTable.Done()
+
+	for _, t := range s.ts {
+		if err := t.Process(s.id, bufTable.Copy()); err != nil {
 			return err
 		}
 	}

--- a/query/stdlib/testing/testing.go
+++ b/query/stdlib/testing/testing.go
@@ -104,11 +104,6 @@ var FluxEndToEndSkipList = map[string]string{
 
 	"window_group_mean_ungroup": "window trigger optimization modifies sort order of its output tables (https://github.com/influxdata/flux/issues/1067)",
 
-	// Flaky test cases, see (https://github.com/influxdata/influxdb/issues/12891)
-	"range":                 "flaky test (https://github.com/influxdata/influxdb/issues/12891)",
-	"window_generate_empty": "flaky test (https://github.com/influxdata/influxdb/issues/12891)",
-	"parse_regex":           "flaky test (https://github.com/influxdata/influxdb/issues/12891)",
-
 	"median_column": "failing in different ways (https://github.com/influxdata/influxdb/issues/13909)",
 	"dynamic_query": "panic when executing",
 }

--- a/storage/reads/table.gen.go
+++ b/storage/reads/table.gen.go
@@ -44,7 +44,6 @@ func newFloatTable(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -79,11 +78,13 @@ func (t *floatTable) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -91,24 +92,23 @@ func (t *floatTable) Do(f func(flux.ColReader) error) error {
 
 func (t *floatTable) Done() {}
 
-func (t *floatTable) advance() bool {
-	for _, cb := range t.colBufs {
-		if cb != nil {
-			cb.Release()
-		}
-	}
-
+func (t *floatTable) advance() *colReader {
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
-		return false
+	l := a.Len()
+	if l == 0 {
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 // group table
@@ -137,7 +137,6 @@ func newFloatGroupTable(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -162,11 +161,13 @@ func (t *floatGroupTable) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -174,23 +175,28 @@ func (t *floatGroupTable) Do(f func(flux.ColReader) error) error {
 
 func (t *floatGroupTable) Done() {}
 
-func (t *floatGroupTable) advance() bool {
+func (t *floatGroupTable) advance() *colReader {
 RETRY:
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
+	l := a.Len()
+	if l == 0 {
 		if t.advanceCursor() {
 			goto RETRY
 		}
 
-		return false
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 func (t *floatGroupTable) advanceCursor() bool {
@@ -253,7 +259,6 @@ func newIntegerTable(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -288,11 +293,13 @@ func (t *integerTable) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -300,24 +307,23 @@ func (t *integerTable) Do(f func(flux.ColReader) error) error {
 
 func (t *integerTable) Done() {}
 
-func (t *integerTable) advance() bool {
-	for _, cb := range t.colBufs {
-		if cb != nil {
-			cb.Release()
-		}
-	}
-
+func (t *integerTable) advance() *colReader {
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
-		return false
+	l := a.Len()
+	if l == 0 {
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 // group table
@@ -346,7 +352,6 @@ func newIntegerGroupTable(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -371,11 +376,13 @@ func (t *integerGroupTable) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -383,23 +390,28 @@ func (t *integerGroupTable) Do(f func(flux.ColReader) error) error {
 
 func (t *integerGroupTable) Done() {}
 
-func (t *integerGroupTable) advance() bool {
+func (t *integerGroupTable) advance() *colReader {
 RETRY:
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
+	l := a.Len()
+	if l == 0 {
 		if t.advanceCursor() {
 			goto RETRY
 		}
 
-		return false
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 func (t *integerGroupTable) advanceCursor() bool {
@@ -462,7 +474,6 @@ func newUnsignedTable(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -497,11 +508,13 @@ func (t *unsignedTable) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -509,24 +522,23 @@ func (t *unsignedTable) Do(f func(flux.ColReader) error) error {
 
 func (t *unsignedTable) Done() {}
 
-func (t *unsignedTable) advance() bool {
-	for _, cb := range t.colBufs {
-		if cb != nil {
-			cb.Release()
-		}
-	}
-
+func (t *unsignedTable) advance() *colReader {
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
-		return false
+	l := a.Len()
+	if l == 0 {
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 // group table
@@ -555,7 +567,6 @@ func newUnsignedGroupTable(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -580,11 +591,13 @@ func (t *unsignedGroupTable) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -592,23 +605,28 @@ func (t *unsignedGroupTable) Do(f func(flux.ColReader) error) error {
 
 func (t *unsignedGroupTable) Done() {}
 
-func (t *unsignedGroupTable) advance() bool {
+func (t *unsignedGroupTable) advance() *colReader {
 RETRY:
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
+	l := a.Len()
+	if l == 0 {
 		if t.advanceCursor() {
 			goto RETRY
 		}
 
-		return false
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 func (t *unsignedGroupTable) advanceCursor() bool {
@@ -671,7 +689,6 @@ func newStringTable(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -706,11 +723,13 @@ func (t *stringTable) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -718,24 +737,23 @@ func (t *stringTable) Do(f func(flux.ColReader) error) error {
 
 func (t *stringTable) Done() {}
 
-func (t *stringTable) advance() bool {
-	for _, cb := range t.colBufs {
-		if cb != nil {
-			cb.Release()
-		}
-	}
-
+func (t *stringTable) advance() *colReader {
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
-		return false
+	l := a.Len()
+	if l == 0 {
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 // group table
@@ -764,7 +782,6 @@ func newStringGroupTable(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -789,11 +806,13 @@ func (t *stringGroupTable) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -801,23 +820,28 @@ func (t *stringGroupTable) Do(f func(flux.ColReader) error) error {
 
 func (t *stringGroupTable) Done() {}
 
-func (t *stringGroupTable) advance() bool {
+func (t *stringGroupTable) advance() *colReader {
 RETRY:
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
+	l := a.Len()
+	if l == 0 {
 		if t.advanceCursor() {
 			goto RETRY
 		}
 
-		return false
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 func (t *stringGroupTable) advanceCursor() bool {
@@ -880,7 +904,6 @@ func newBooleanTable(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -915,11 +938,13 @@ func (t *booleanTable) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -927,24 +952,23 @@ func (t *booleanTable) Do(f func(flux.ColReader) error) error {
 
 func (t *booleanTable) Done() {}
 
-func (t *booleanTable) advance() bool {
-	for _, cb := range t.colBufs {
-		if cb != nil {
-			cb.Release()
-		}
-	}
-
+func (t *booleanTable) advance() *colReader {
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
-		return false
+	l := a.Len()
+	if l == 0 {
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 // group table
@@ -973,7 +997,6 @@ func newBooleanGroupTable(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -998,11 +1021,13 @@ func (t *booleanGroupTable) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -1010,23 +1035,28 @@ func (t *booleanGroupTable) Do(f func(flux.ColReader) error) error {
 
 func (t *booleanGroupTable) Done() {}
 
-func (t *booleanGroupTable) advance() bool {
+func (t *booleanGroupTable) advance() *colReader {
 RETRY:
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
+	l := a.Len()
+	if l == 0 {
 		if t.advanceCursor() {
 			goto RETRY
 		}
 
-		return false
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 func (t *booleanGroupTable) advanceCursor() bool {

--- a/storage/reads/table.gen.go.tmpl
+++ b/storage/reads/table.gen.go.tmpl
@@ -38,7 +38,6 @@ func new{{.Name}}Table(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -73,11 +72,13 @@ func (t *{{.name}}Table) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil  {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -85,24 +86,23 @@ func (t *{{.name}}Table) Do(f func(flux.ColReader) error) error {
 
 func (t *{{.name}}Table) Done() {}
 
-func (t *{{.name}}Table) advance() bool {
-	for _, cb := range t.colBufs {
-		if cb != nil {
-			cb.Release()
-		}
-	}
-
+func (t *{{.name}}Table) advance() *colReader {
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
-		return false
+	l := a.Len()
+	if l == 0 {
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 // group table
@@ -131,7 +131,6 @@ func new{{.Name}}GroupTable(
 		cur:   cur,
 	}
 	t.readTags(tags)
-	t.advance()
 
 	return t
 }
@@ -156,11 +155,13 @@ func (t *{{.name}}GroupTable) Do(f func(flux.ColReader) error) error {
 		t.mu.Unlock()
 	}()
 
-	if !t.Empty() {
-		t.err = f(t)
-		for !t.isCancelled() && t.err == nil && t.advance() {
-			t.err = f(t)
+	for !t.isCancelled() && t.err == nil  {
+		cr := t.advance()
+		if cr == nil {
+			break
 		}
+		t.err = f(cr)
+		cr.Release()
 	}
 
 	return t.err
@@ -168,23 +169,28 @@ func (t *{{.name}}GroupTable) Do(f func(flux.ColReader) error) error {
 
 func (t *{{.name}}GroupTable) Done() {}
 
-func (t *{{.name}}GroupTable) advance() bool {
+func (t *{{.name}}GroupTable) advance() *colReader {
 RETRY:
 	a := t.cur.Next()
-	t.l = a.Len()
-	if t.l == 0 {
+	l := a.Len()
+	if l == 0 {
 		if t.advanceCursor() {
 			goto RETRY
 		}
 
-		return false
+		return nil
 	}
 
-	t.colBufs[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
-	t.colBufs[valueColIdx] = t.toArrowBuffer(a.Values)
-	t.appendTags()
-	t.appendBounds()
-	return true
+	// Retrieve the buffer for the data to avoid allocating
+	// additional slices. If the buffer is still being used
+	// because the references were retained, then we will
+	// allocate a new buffer.
+	cr := t.getBuffer(l)
+	cr.cols[timeColIdx] = arrow.NewInt(a.Timestamps, t.alloc)
+	cr.cols[valueColIdx] = t.toArrowBuffer(a.Values)
+	t.appendTags(cr)
+	t.appendBounds(cr)
+	return cr
 }
 
 func (t *{{.name}}GroupTable) advanceCursor() bool {

--- a/storage/reads/table.go
+++ b/storage/reads/table.go
@@ -25,10 +25,7 @@ type table struct {
 
 	done chan struct{}
 
-	// The current number of records in memory
-	l int
-
-	colBufs []array.Interface
+	colBufs *colReader
 
 	err error
 
@@ -45,22 +42,20 @@ func newTable(
 	alloc *memory.Allocator,
 ) table {
 	return table{
-		done:    done,
-		bounds:  bounds,
-		key:     key,
-		tags:    make([][]byte, len(cols)),
-		defs:    defs,
-		colBufs: make([]array.Interface, len(cols)),
-		cols:    cols,
-		alloc:   alloc,
+		done:   done,
+		bounds: bounds,
+		key:    key,
+		tags:   make([][]byte, len(cols)),
+		defs:   defs,
+		cols:   cols,
+		alloc:  alloc,
 	}
 }
 
 func (t *table) Key() flux.GroupKey   { return t.key }
 func (t *table) Cols() []flux.ColMeta { return t.cols }
 func (t *table) Err() error           { return t.err }
-func (t *table) Empty() bool          { return t.l == 0 }
-func (t *table) Len() int             { return t.l }
+func (t *table) Empty() bool          { return false }
 
 func (t *table) Cancel() {
 	atomic.StoreInt32(&t.cancelled, 1)
@@ -70,37 +65,78 @@ func (t *table) isCancelled() bool {
 	return atomic.LoadInt32(&t.cancelled) != 0
 }
 
-func (t *table) Retain()  {}
-func (t *table) Release() {}
+// getBuffer will retrieve a suitable buffer for the
+// table implementations to use. If the existing buffer
+// is not used anymore, then it may be reused.
+func (t *table) getBuffer(l int) *colReader {
+	if t.colBufs != nil && atomic.LoadInt64(&t.colBufs.refCount) == 0 {
+		return t.colBufs
+	}
 
-func (t *table) Bools(j int) *array.Boolean {
-	execute.CheckColType(t.cols[j], flux.TBool)
-	return t.colBufs[j].(*array.Boolean)
+	// The current buffer is still being used so we should
+	// generate a new one.
+	t.colBufs = &colReader{
+		refCount: 1,
+		key:      t.key,
+		colMeta:  t.cols,
+		cols:     make([]array.Interface, len(t.cols)),
+		l:        l,
+	}
+	return t.colBufs
 }
 
-func (t *table) Ints(j int) *array.Int64 {
-	execute.CheckColType(t.cols[j], flux.TInt)
-	return t.colBufs[j].(*array.Int64)
+type colReader struct {
+	refCount int64
+
+	key     flux.GroupKey
+	colMeta []flux.ColMeta
+	cols    []array.Interface
+	l       int
 }
 
-func (t *table) UInts(j int) *array.Uint64 {
-	execute.CheckColType(t.cols[j], flux.TUInt)
-	return t.colBufs[j].(*array.Uint64)
+func (cr *colReader) Retain() {
+	atomic.AddInt64(&cr.refCount, 1)
+}
+func (cr *colReader) Release() {
+	if atomic.AddInt64(&cr.refCount, -1) == 0 {
+		for _, col := range cr.cols {
+			col.Release()
+		}
+	}
 }
 
-func (t *table) Floats(j int) *array.Float64 {
-	execute.CheckColType(t.cols[j], flux.TFloat)
-	return t.colBufs[j].(*array.Float64)
+func (cr *colReader) Key() flux.GroupKey   { return cr.key }
+func (cr *colReader) Cols() []flux.ColMeta { return cr.colMeta }
+func (cr *colReader) Len() int             { return cr.l }
+
+func (cr *colReader) Bools(j int) *array.Boolean {
+	execute.CheckColType(cr.colMeta[j], flux.TBool)
+	return cr.cols[j].(*array.Boolean)
 }
 
-func (t *table) Strings(j int) *array.Binary {
-	execute.CheckColType(t.cols[j], flux.TString)
-	return t.colBufs[j].(*array.Binary)
+func (cr *colReader) Ints(j int) *array.Int64 {
+	execute.CheckColType(cr.colMeta[j], flux.TInt)
+	return cr.cols[j].(*array.Int64)
 }
 
-func (t *table) Times(j int) *array.Int64 {
-	execute.CheckColType(t.cols[j], flux.TTime)
-	return t.colBufs[j].(*array.Int64)
+func (cr *colReader) UInts(j int) *array.Uint64 {
+	execute.CheckColType(cr.colMeta[j], flux.TUInt)
+	return cr.cols[j].(*array.Uint64)
+}
+
+func (cr *colReader) Floats(j int) *array.Float64 {
+	execute.CheckColType(cr.colMeta[j], flux.TFloat)
+	return cr.cols[j].(*array.Float64)
+}
+
+func (cr *colReader) Strings(j int) *array.Binary {
+	execute.CheckColType(cr.colMeta[j], flux.TString)
+	return cr.cols[j].(*array.Binary)
+}
+
+func (cr *colReader) Times(j int) *array.Int64 {
+	execute.CheckColType(cr.colMeta[j], flux.TTime)
+	return cr.cols[j].(*array.Int64)
 }
 
 // readTags populates b.tags with the provided tags
@@ -120,32 +156,32 @@ func (t *table) readTags(tags models.Tags) {
 }
 
 // appendTags fills the colBufs for the tag columns with the tag value.
-func (t *table) appendTags() {
+func (t *table) appendTags(cr *colReader) {
 	for j := range t.cols {
 		v := t.tags[j]
 		if v != nil {
 			b := arrow.NewStringBuilder(t.alloc)
-			b.Reserve(t.l)
-			b.ReserveData(t.l * len(v))
-			for i := 0; i < t.l; i++ {
+			b.Reserve(cr.l)
+			b.ReserveData(cr.l * len(v))
+			for i := 0; i < cr.l; i++ {
 				b.Append(v)
 			}
-			t.colBufs[j] = b.NewArray()
+			cr.cols[j] = b.NewArray()
 			b.Release()
 		}
 	}
 }
 
 // appendBounds fills the colBufs for the time bounds
-func (t *table) appendBounds() {
+func (t *table) appendBounds(cr *colReader) {
 	bounds := []execute.Time{t.bounds.Start, t.bounds.Stop}
 	for j := range []int{startColIdx, stopColIdx} {
 		b := arrow.NewIntBuilder(t.alloc)
-		b.Reserve(t.l)
-		for i := 0; i < t.l; i++ {
+		b.Reserve(cr.l)
+		for i := 0; i < cr.l; i++ {
 			b.UnsafeAppend(int64(bounds[j]))
 		}
-		t.colBufs[j] = b.NewArray()
+		cr.cols[j] = b.NewArray()
 		b.Release()
 	}
 }

